### ISSUE TITLE
ftbfs: gcc 14+: fail to compile

### ISSUE THREAD

**Issue by powerman** [clan/runit#2]
ftbfs branch commit a9089b1 fail to compile on gcc 14:

```
./compile prot.c
prot.c: In function ‘prot_gid’:
prot.c:17:19: error: passing argument 2 of ‘setgroups’ from incompatible pointer type [-Wincompatible-pointer-types]
   17 |   if (setgroups(1,x) == -1) return -1;
      |                   ^
      |                   |
      |                   short int *
In file included from prot.c:4:
/usr/include/grp.h:180:50: note: expected ‘const __gid_t *’ {aka ‘const unsigned int *’} but argument is of type ‘short int *’
  180 | extern int setgroups (size_t __n, const __gid_t *__groups) __THROW;
      |                                   ~~~~~~~~~~~~~~~^~~~~~~~
make: *** [Makefile:296: prot.o] Error 1
```

gcc (Gentoo Hardened 14.2.1_p20240817 p4) 14.2.1 20240817

**Comment by powerman** [comment]
Same error with gcc (Gentoo Hardened 15.0.0_pre20240818-r1 p10) 15.0.0 20240818 (experimental).

**Comment by clan** [comment]
Get, this might be fixed in my latest patch. I'll reorganize the patch in following order:

1. remove some redirect to >/dev/null, otherwise some unexpected error is hidden, then the feature test give unexpected result for later compilation
2. fix all the unexpected error in feature test code
3. fix all the compilation error(s) in build, ignore all warnings
4. fix compilation warning(s), ordered by warning type, debian's patch first (with debian's url mentioned)
5. fix remain compilation warning(s) 
6. add new agreed new feature (if any)

after 1-4 done, I'll prepare the ebuild patch for PR.

BTW: I'll do git push --force for FTBFS branch before the ebuild patch is merged.

**Comment by clan** [comment]
thanks for the explanation, I'm just wondering why chkshsgr.c of 0014-fix-implicit-declarations-incompatible-types-argv-ty.patch change short x[4] to gid_t x[4], while tryshsgr.c 0029-fix-ftbfs-with-gcc14.patch do the cast instead. I'll review this again and fix it. So should both chkshsgr.c & tryshsgr.c keep using 'short x[4]' while use cast to elimate the compiler warning?

BTW, do you have any plan to put a official github repo for runit? So we (both debian, gentoo, ...) can submit patch related to the software itself and have them merged in a single place, while keep distribution's build patch alone.

**Comment by clan** [comment]
@powerman I just update the gentoo portage (https://github.com/clan/gentoo/tree/runit/sys-process/runit) w/ 6 patches applied (without warning fix patch), please have a test whenver you have time. Let me know if there are any problems, thanks.

**Comment by powerman** [comment]
Sure. I'll check it tomorrow.

**Comment by powerman** [comment]
@clan 
- While I'm glad to see removed `RDEPEND="sys-apps/openrc"` it's not that simple. The actual dependency is in `files/{1,3}-2.1.2` - they call `/sbin/openrc`. To remove `RDEPEND` we need to fix these files first. Maybe even drop these files at all (i.e. add a USE-flag which will either install openrc-based version of these files or nothing). As it is complicated I propose to avoid making this (`RDEPEND`) change in first PR at all (i.e. add it back to the ebuild).
- There are also other changes in the ebuild to be discussed. I don't think continue such discussions in unrelated issues (like this one) is a good idea, maybe you'll enable GitHub Discussions for this repo and we'll move there?
- `0005-support-etc-runit-nosync-file-to-make-sync-on-shutdo.patch` - I don't think we need it. Sure, it was applied by @g-pape, which is a good sign. But it's related to decade-old issue someone had with Debian vserver for actually unknown root reason. AFAIK Gentoo as of today does not support vserver at all, so I vote against adding such a useless feature.
- `0006-add-update-doc-man-page-of-chpst-and-runsv.patch`:
  - doc updates are nice to have, but I'm not sure why do we need this patch in first PR
  - some doc use weird formatting, which probably should be fixed (e.g. patch lines 107-108 contains word hyphenation, which is completely useless in html, also there are a lot of weird spacing, too long lines which makes patch harder to read for no reason, etc.)

**Comment by powerman** [comment]
With these 6 patches it still fails to compile on:
- gcc (Gentoo Hardened 14.2.1_p20240817 p4) 14.2.1 20240817
- gcc (Gentoo Hardened 15.0.0_pre20240818-r1 p10) 15.0.0 20240818 (experimental)

```
./compile chkshsgr.c
chkshsgr.c: In function 'main':
chkshsgr.c:11:19: error: passing argument 2 of 'getgroups' from incompatible pointer type [-Wincompatible-pointer-types]
   11 |   if (getgroups(1,x) == 0) if (setgroups(1,x) == -1) _exit(1);
      |                   ^
      |                   |
      |                   short int *
In file included from /usr/include/features.h:503,
                 from /usr/include/unistd.h:25,
                 from chkshsgr.c:3:
/usr/include/bits/unistd.h:113:1: note: expected '__gid_t *' {aka 'unsigned int *'} but argument is of type 'short int *'
  113 | __NTH (getgroups (int __size, __gid_t __list[]))
      | ^~~~~
chkshsgr.c:11:44: error: passing argument 2 of 'setgroups' from incompatible pointer type [-Wincompatible-pointer-types]
   11 |   if (getgroups(1,x) == 0) if (setgroups(1,x) == -1) _exit(1);
      |                                            ^
      |                                            |
      |                                            short int *
In file included from chkshsgr.c:4:
/usr/include/grp.h:180:50: note: expected 'const __gid_t *' {aka 'const unsigned int *'} but argument is of type 'short int *'
  180 | extern int setgroups (size_t __n, const __gid_t *__groups) __THROW;
      |                                   ~~~~~~~~~~~~~~~^~~~~~~~
make: *** [Makefile:177: chkshsgr.o] Error 1
```

Seriously, let's try to fix gcc/clang support for all available in Gentoo versions first, and then in following PRs work on other patches like doc updates etc.

**Comment by clan** [comment]
> ./compile chkshsgr.c
> chkshsgr.c: In function 'main':
> chkshsgr.c:11:19: error: passing argument 2 of 'getgroups' from incompatible pointer type [-Wincompatible-pointer-types]

It looks like "-Werror" is in CFLAGS, if not, can you post the full compile command (edit src/compile, add -x for sh)?

**Comment by clan** [comment]
> * While I'm glad to see removed `RDEPEND="sys-apps/openrc"` it's not that simple. The actual dependency is in `files/{1,3}-2.1.2` - they call `/sbin/openrc`. To remove `RDEPEND` we need to fix these files first. Maybe even drop these files at all (i.e. add a USE-flag which will either install openrc-based version of these files or nothing). As it is complicated I propose to avoid making this (`RDEPEND`) change in first PR at all (i.e. add it back to the ebuild).
> * There are also other changes in the ebuild to be discussed. I don't think continue such discussions in unrelated issues (like this one) is a good idea, maybe you'll enable GitHub Discussions for this repo and we'll move there?

OK. I know all these issues. I'd like to keep these change unchanged (same to -r4 ebuild) for now. Let's discuss gentoo specific issue later in other place.

> * `0005-support-etc-runit-nosync-file-to-make-sync-on-shutdo.patch` - I don't think we need it. Sure, it was applied by @g-pape, which is a good sign. But it's related to decade-old issue someone had with Debian vserver for actually unknown root reason. AFAIK Gentoo as of today does not support vserver at all, so I vote against adding such a useless feature.

I don't think this vserver is debian specific only. The original [debian bug](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=695281) said: This is desirable on physical hosts, of course, but not in a vserver (which is basically just a glorified chroot).
It might be useful for other chroot purpose, or docker (which is another kind of chroot).

> * `0006-add-update-doc-man-page-of-chpst-and-runsv.patch`:
>   
>   * doc updates are nice to have, but I'm not sure why do we need this patch in first PR
>   * some doc use weird formatting, which probably should be fixed (e.g. patch lines 107-108 contains word hyphenation, which is completely useless in html, also there are a lot of weird spacing, too long lines which makes patch harder to read for no reason, etc.)

This patch add some important missing information (explantion of chpst -t option, ...), user won't know without read the source.
I'll fix the text issue in line 107 and other place(s) if found.

**Comment by powerman** [comment]
> It looks like "-Werror" is in CFLAGS, if not, can you post the full compile command (edit src/compile, add -x for sh)?

No, it's not:

```
$ pwd
/home/powerman/proj/soft/runit/admin/runit-2.1.2/compile
$ env | grep FLAGS
$ gcc --version | head -n 1
gcc (Gentoo Hardened 14.2.1_p20240817 p4) 14.2.1 20240817
$ gcc -O2 -Wall -c chkshsgr.c
chkshsgr.c: In function 'main':
chkshsgr.c:11:19: error: passing argument 2 of 'getgroups' from incompatible pointer type [-Wincompatible-pointer-types]
   11 |   if (getgroups(1,x) == 0) if (setgroups(1,x) == -1) _exit(1);
      |                   ^
      |                   |
      |                   short int *
In file included from /usr/include/features.h:503,
                 from /usr/include/unistd.h:25,
                 from chkshsgr.c:3:
/usr/include/bits/unistd.h:113:1: note: expected '__gid_t *' {aka 'unsigned int *'} but argument is of type 'short int *'
  113 | __NTH (getgroups (int __size, __gid_t __list[]))
      | ^~~~~
chkshsgr.c:11:44: error: passing argument 2 of 'setgroups' from incompatible pointer type [-Wincompatible-pointer-types]
   11 |   if (getgroups(1,x) == 0) if (setgroups(1,x) == -1) _exit(1);
      |                                            ^
      |                                            |
      |                                            short int *
In file included from chkshsgr.c:4:
/usr/include/grp.h:180:50: note: expected 'const __gid_t *' {aka 'const unsigned int *'} but argument is of type 'short int *'
  180 | extern int setgroups (size_t __n, const __gid_t *__groups) __THROW;
      |                                   ~~~~~~~~~~~~~~~^~~~~~~~
```

> It might be useful for other chroot purpose, or docker (which is another kind of chroot).

It might. Or it might not. Until there are any Gentoo bugreports there is no issue to fix.
It's not a package maintainer's work to add features no one needs - it's a package author's work.
Package maintainer might add feature requested by distribution users if package author does not wanna add it - I suppose @g-pape acted as a package maintainer while adding this feature as a Debian patch as a way to fix a Debian bugreport. If we were had similar bug open in Gentoo we probably have to do the same (after trying to find out a root reason first, of course).
But at this moment and for the PR fixing runit compiling issues this feature is absolutely redundant.
If you disagree about usefulness of the feature itself then just add it in one of following PRs, we'll both explain our points in that PR's comments and let Gentoo devs decide to merge it or not.

> This patch add some important missing information (explantion of chpst -t option, ...), user won't know without read the source.

I'm 100% agree this patch is needed, I just wanna avoid too large PR with unrelated patches because it was the reason why previous PR was rejected. Let's make first PR only with patches which fix compilation issues (plus maybe extra `/dev/null` patch because it's related to work on compilation issues) and nothing else. After it'll be merged let's open more small PRs with separate patches - one for docs, another for "nosync" feature, few more for fixing compiler warnings, etc.

**Comment by clan** [comment]
> No, it's not:
> 
> ```
> $ gcc --version | head -n 1

OK, I'll build gcc 14 & 15 to test and try to find reason. I'm using gcc 13 & clang 18 for the build test.

> > It might be useful for other chroot purpose, or docker (which is another kind of chroot).
> 
> It might. Or it might not. Until there are any Gentoo bugreports there is no issue to fix. It's not a package maintainer's work to add features no one needs - it's a package author's work. Package maintainer might add feature requested by distribution users if package author does not wanna add it - I suppose @g-pape acted as a package maintainer while adding this feature as a Debian patch as a way to fix a Debian bugreport. If we were had similar bug open in Gentoo we probably have to do the same (after trying to find out a root reason first, of course). But at this moment and for the PR fixing runit compiling issues this feature is absolutely redundant. If you disagree about usefulness of the feature itself then just add it in one of following PRs, we'll both explain our points in that PR's comments and let Gentoo devs decide to merge it or not.

This one add a new feature (which is useful in chroot or docker if system have high IO cause sync very slowly, I know this issue for a long time), the code is very clear which is unlikely to cause problem, and it's provided by the author. I don't see any reason to not add this patch. And sure, I can add this patch as an option and let Gentoo devs do the final decision.

> 
> > This patch add some important missing information (explantion of chpst -t option, ...), user won't know without read the source.
> 
> I'm 100% agree this patch is needed, I just wanna avoid too large PR with unrelated patches because it was the reason why previous PR was rejected. Let's make first PR only with patches which fix compilation issues (plus maybe extra `/dev/null` patch because it's related to work on compilation issues) and nothing else. After it'll be merged let's open more small PRs with separate patches - one for docs, another for "nosync" feature, few more for fixing compiler warnings, etc.

The documentation change is unlikely to cause code problem (with only two small change in the .c file), so it's unlikely to cause problem and easily to be reviewd. And again, let Gentoo devs do the final decision.

I'd asked Gerrit Pape about the possiblity of maintain runit on github, but it seems that won't happen soon. Maybe we should discuss this in the maillist?

**Comment by powerman** [comment]
> I'd asked Gerrit Pape about the possiblity of maintain runit on github, but it seems that won't happen soon. Maybe we should discuss this in the maillist?

It already was discussed there several time. You're free to start another one discussion, of course, but I think it makes no sense - Gerrit knows everyone wants this, but it's up to him to do this or not.

**Comment by clan** [comment]
@powerman ignore last reply. I do reproduce the error, it seems gcc 14 turn -Wincompatible-pointer-types (and many more, see https://gcc.gnu.org/gcc-14/porting_to.html) into error, I'll check why and fix them.

BTW it seems it might be valulable for the rest of the warning fix to be included, :-(

**Comment by clan** [comment]
I've update the patch on branch ftbfs.v2 (https://github.com/clan/runit/tree/ftbfs.v2), fix all the errors and most of warnings, tested by using clang 18 and gcc 15.

gentoo portage have been force pushed to include the patch for errors (warnings before gcc 14) by using gcc 14, please test and confirm. If no more problem, I'll submit PR (w/ the options explained) on gentoo for final decision. thanks. @powerman please test when have time, thanks.

**Comment by powerman** [comment]
I confirm it builds cleanly with gcc 11…15.

It looks like you insist on including unrelated to gcc patches (docs update and nosync feature) in first PR. It means we'll have to continue this discussion in the PR - something which I'd like to avoid to ensure gcc patches will be merged without any extra issues. But it's up to you how to handle this.

Also I'm unsure is commit 12e2558af914769bae567da35c073d9664e1b058 valid. As @g-pape said in https://github.com/clan/runit/commit/5a7b8099ae638ce72abae2c4d8cac7ce33477522#r145939945 the usage of `short` type is intentional. And while you didn't change the type definition you've anyway converted all usages of the var to `gid_t` - which to me looks the same as changing type in var's definition (but I'm not a C expert and may be wrong about this). Also this commit's description message is huge and contains a lot of unrelated to this change info.

Doc patch has added some `<dt>` and `<dd>` tags into html, but it missed `<dl>` tag which should surround these. Not sure is this critical issue and is html was auto-generated from man page or crafted manually.

Commit ee1867e081d42d7231df41696efb00d6f4624673 has removed `register` from several function's params. At a glance commit description does not mention why this needed. I guess `register` is a performance hint for a compiler, so probably this isn't critical, but if this change wasn't intentional it may worth to be undone and if it was intentional it may worth to explain reasoning in commit description.

Some added `/* fallthrough */` looks invalid - in cases when they follow `*usage()` or `done()` calls, which I suppose will call `exit()`. It is harmless, but misleading and thus may worth to be removed or (if they needed to silence some compiler warning) changed to e.g. `/* exited, so no fallthrough */`.

While commit 486c348c83512968f2eba6390130e7125053e1cd fixes the warning, they way it's done looks really weird: `static char hex[17] ="0123456789abcdef";`. Maybe there are some other ways to fix this? At a glance `static char hex[] ="0123456789abcdef";` should work too - at least there is no warning in gcc 15.

It looks like I've reviewed just about a half of the changes so far. So, to be continued…

**Comment by powerman** [comment]
Commit 5ca8e7d25782fe79ac364de818e6f681fe5e5b1b added a lot of `// TODO` comments in places of "missing" error handling. At a glance I can't say is ignoring error from `write()` to a pipe is correct, but all other places (`fchdir` and `ftruncate`) looks correct. To me it looks like these warnings are false positives and "fixing" them makes code worse - you've added useless checks for returned value (I hope compiler will optimize them away). If there are no better way to silence these warnings without adding useless `if` with some body then I propose to change `// TODO` comment to `// Ignored.` comment, because TODO is misleading and creates impression these errors actually needs some handling, which is not the case (well, except I'm not sure about errors from writes to a pipe).

Commit ee1867e081d42d7231df41696efb00d6f4624673 has commented out `extern int select();` in `admin/runit-2.1.2/src/select.h{1,2}` - why not just remove this line instead?

Commit dc7dfcb88a98d45d905ae6cc23d64db9dff06bcc - is the only way to mark variable "used" in C is to `printf()` it? I suppose there are should be other ways, without side effects.

Commit 531c5e1a948767235d02bb3b352119bba9ed950e changes `warn-auto.sh`, which is "auto-generated. Do not edit!". At a glance it looks like leftover after some manual debugging. If this change is actually needed then it probably should be done in place which generates that file.

**Comment by powerman** [comment]
So, looks like there are a lot of things to clean up (if you like to), but in general everything looks good to me (except a few places where my expertise is not enough and thus I'm unsure). Great work, thanks!

**Comment by clan** [comment]
> I confirm it builds cleanly with gcc 11…15.

thanks, I'll prepare the PR for gentoo w/ all the patches. I'll describe each of the the patches:
1. required, if not build will failed
2. optional, but why it's ok to be added in.

Because there is no active upstream for runit, I don't known how the next step will evolve. But in general, these patches just fix the compiler errors and most of warnings, add nothing new except the one from the author. So I won't insist to have all theese patches in Gentoo.

> Also I'm unsure is commit [12e2558](https://github.com/clan/runit/commit/12e2558af914769bae567da35c073d9664e1b058) valid. As @g-pape said in [5a7b809#r145939945](https://github.com/clan/runit/commit/5a7b8099ae638ce72abae2c4d8cac7ce33477522#r145939945) the usage of `short` type is intentional. And while you didn't change the type definition you've anyway converted all usages of the var to `gid_t` - which to me looks the same as changing type in var's definition (but I'm not a C expert and may be wrong about this). Also this commit's description message is huge and contains a lot of unrelated to this change info.

the cast in (get|set)group is ok and must if compiled w/ gcc 14 or above. The debian's patch do the same thing.
 
> Doc patch has added some `<dt>` and `<dd>` tags into html, but it missed `<dl>` tag which should surround these. Not sure is this critical issue and is html was auto-generated from man page or crafted manually.

I'll fix this. But I'll only focus on the changed part. HTML doesn't require strict match on tags.
 
> Commit [ee1867e](https://github.com/clan/runit/commit/ee1867e081d42d7231df41696efb00d6f4624673) has removed `register` from several function's params. At a glance commit description does not mention why this needed. I guess `register` is a performance hint for a compiler, so probably this isn't critical, but if this change wasn't intentional it may worth to be undone and if it was intentional it may worth to explain reasoning in commit description.

I think modern compiler can handle this better. In fact, I haven't seen the use of the register keyword longtime, I'll do more research to get a better reason, if not, it's ok to still keep the keyword as is.

> Some added `/* fallthrough */` looks invalid - in cases when they follow `*usage()` or `done()` calls, which I suppose will call `exit()`. It is harmless, but misleading and thus may worth to be removed or (if they needed to silence some compiler warning) changed to e.g. `/* exited, so no fallthrough */`.

The comment here is just a hint to silence the compiler warnings. it just mean execution will continue to next case or default because no break in this case. It doesn't mean exit. Since it's a comment, it add nothing in fact.
 
> While commit [486c348](https://github.com/clan/runit/commit/486c348c83512968f2eba6390130e7125053e1cd) fixes the warning, they way it's done looks really weird: `static char hex[17] ="0123456789abcdef";`. Maybe there are some other ways to fix this? At a glance `static char hex[] ="0123456789abcdef";` should work too - at least there is no warning in gcc 15.

The following code will access the character in hex w/ careful design (& 15) to avoid out of range access. So it's still ok if don't change, but in C, string always terminated with \0, so I change 16 to 17 to silence the compiler warning.
 
> It looks like I've reviewed just about a half of the changes so far. So, to be continued…

**Comment by clan** [comment]
> Commit [5ca8e7d](https://github.com/clan/runit/commit/5ca8e7d25782fe79ac364de818e6f681fe5e5b1b) added a lot of `// TODO` comments in places of "missing" error handling. At a glance I can't say is ignoring error from `write()` to a pipe is correct, but all other places (`fchdir` and `ftruncate`) looks correct. To me it looks like these warnings are false positives and "fixing" them makes code worse - you've added useless checks for returned value (I hope compiler will optimize them away). If there are no better way to silence these warnings without adding useless `if` with some body then I propose to change `// TODO` comment to `// Ignored.` comment, because TODO is misleading and creates impression these errors actually needs some handling, which is not the case (well, except I'm not sure about errors from writes to a pipe).

I intend to handle these return value check (do log etc.), but I'm not sure what is the best to do until understanding the whole logic, so I left a TODO here, which is a prompt for later work (if there will).

> Commit [ee1867e](https://github.com/clan/runit/commit/ee1867e081d42d7231df41696efb00d6f4624673) has commented out `extern int select();` in `admin/runit-2.1.2/src/select.h{1,2}` - why not just remove this line instead?

The comment is to keep a link with the file name. In fact, we can remove select.h[12].
 
> Commit [dc7dfcb](https://github.com/clan/runit/commit/dc7dfcb88a98d45d905ae6cc23d64db9dff06bcc) - is the only way to mark variable "used" in C is to `printf()` it? I suppose there are should be other ways, without side effects.

These three files are all part of the feature test code and won't be linked into the final executables. We can use __attrribute__ for the same purpose, but it might depend on compiler, so I choose a generic way.
 
> Commit [531c5e1](https://github.com/clan/runit/commit/531c5e1a948767235d02bb3b352119bba9ed950e) changes `warn-auto.sh`, which is "auto-generated. Do not edit!". At a glance it looks like leftover after some manual debugging. If this change is actually needed then it probably should be done in place which generates that file.

This file is a template to generate other machine-generated file, so the text is for the generated file, not for itself.

**Comment by powerman** [comment]
> Because there is no active upstream for runit, I don't known how the next step will evolve.

As you'll be a proxy-maintainer for the package, next step will be to handle new bugreports. Sometimes it may result in need to craft some extra patch - just like @g-pape did for "nosync" feature. Of course, nothing stops you from making other changes if you like to, e.g. improve ebuild itself or `/etc/runit/{1,2,3}` scripts etc. But in general it's enough to just make sure it can be installed on current Gentoo and handle bugreports.

> I think modern compiler can handle this better.

Maybe. But it's not guaranteed, and optimizing whole source code for modern compilers should be done in a separate patch, with some benchmarking. People are proven to be really bad at estimating which code will work faster, so making such changes without benchmarks to me look like any other _random_ (but safe) change for no reason.

> The comment here is just a hint to silence the compiler warnings.

Then write in a comment "silence the compiler warning" instead of "fallthrough". Right now some of these comments are misleading, which is bad IMO and can be easily fixed in a way I proposed or any other way you like.

> The following code will access the character in hex w/ careful design (& 15) to avoid out of range access. So it's still ok if don't change, but in C, string always terminated with \0, so I change 16 to 17 to silence the compiler warning.

I understood the reasoning but it still looks weird, and if there are other ways to fix this which won't look weird I think it's better to use one these other ways.

> I intend to handle these return value check (do log etc.)

I think this is very likely a bad idea. I suppose there are legal use cases why these operations may fail and thus any extra log output or other "handling" will be wrong and will make these legal use cases looks like something is wrong. For example, in `fchdir` case, the target directory might not exists and that moment - and it's okay.

**Comment by clan** [comment]
> > The comment here is just a hint to silence the compiler warnings.
> 
> Then write in a comment "silence the compiler warning" instead of "fallthrough". Right now some of these comments are misleading, which is bad IMO and can be easily fixed in a way I proposed or any other way you like.

the warning message ( -Wimplicit-fallthrough) already says there might be an implicit fallthrough problem here (because no break like ususal case), so yes, I think no break mean fallthrough to next case, so fallthrough is the right word here. 

> > I intend to handle these return value check (do log etc.)
> 
> I think this is very likely a bad idea. I suppose there are legal use cases why these operations may fail and thus any extra log output or other "handling" will be wrong and will make these legal use cases looks like something is wrong. For example, in `fchdir` case, the target directory might not exists and that moment - and it's okay.

Log might means show error message to stderr. This won't happen soon, and it must be handled with detail use case.

**Comment by clan** [comment]
@powerman 

I'd like to submit PR on gentoo include all the patches except the one w/ new feature (cat. 3) since you've reviewed all the patches. Do you agree?

I'll reorganize the git tree (force push the master branch based on the author's tree w/ release files removed).

1. helpful for development & debug, e.g, provide full command line and result, to check what's going on
2. required, build will failed if not included
3. patch from author w/ new feature provided (don't change existing behaviour)
4. documentation change: w/ tiny (unlikely to cause problem) modification of source code
5. source code changed to silence compiler warnings
6. source code changed to handle obsolete C function

**Comment by powerman** [comment]
> I'd like to submit PR on gentoo include all the patches except the one w/ new feature (cat. 3) since you've reviewed all the patches. Do you agree?

I agree to have all mentioned patches applied, but not in a single PR. Both me and @thesamesam (in https://github.com/gentoo/gentoo/pull/38266#issuecomment-2328142441 just an hour ago) have told you to add only minimal required patches in a first PR, but you anyway wanna add all of them at once. What's for? Why it's so important for you to make just one PR instead of several ones?

As Sam said:

> You have a branch with your patches and you cut a new tag with the patches that are required to make runit build and don't change any "user-observable functionality"

To me it means just:

> 2. required, build will failed if not included

So, I propose to make these PRs, one by one, opening next one only after merging or rejecting a previous one:

1. required, build will failed if not included
2. documentation change: w/ tiny (unlikely to cause problem) modification of source code
3. source code changed to handle obsolete C function
4. source code changed to silence compiler warnings
5. helpful for development & debug, e.g, provide full command line and result, to check what's going on

My rationale for this order is:

1. required in any case
2. sync doc with actual runit features/behaviour, so really important to have
3. chances are obsolete C functions may stop working in the future, so chances are it'll become required anyway
4. fixing all warnings doesn't sounds like a critical issue, but as this work was already done then it's probably good idea to merge it
5. we already discussed these, to me it looks doubtful these changes will be really useful for anyone except you, but you right - there is a chance someone will make a bugreport with a build.log and you'll see something useful in that log because of this change.

I believe you should give Gentoo devs a chance to review these patches in small hunks dedicated to a single clean reason - that's why I think it's better to have 5 PRs instead of just 1.

**Comment by clan** [comment]
> I agree to have all mentioned patches applied, but not in a single PR. Both me and @thesamesam (in [gentoo/gentoo#38266 (comment)](https://github.com/gentoo/gentoo/pull/38266#issuecomment-2328142441) just an hour ago) have told you to add only minimal required patches in a first PR, but you anyway wanna add all of them at once. What's for? Why it's so important for you to make just one PR instead of several ones?

The reason is that the patch is by error(s) or warning(s) driven, it's really simple, nothing complicated, and it don't change any "user-observable functionality".

> So, I propose to make these PRs, one by one, opening next one only after merging or rejecting a previous one:

The same reason here, I do understand your concerns. But I don't see any reason to submit simple patches in multiple step, which is time consuming and not worthwhile. Of course, if gentoo devs think any non-critical patch is too complex, I can remove it for later. Part of the reason might be: no active upstream to merge this patches in official, :-(

> 5. helpful for development & debug, e.g, provide full command line and result, to check what's going on

this one is especially helpful for gentoo, if any bug(s) found, users (some are not developer) can submit the build.log with all the necessary information in.

**Comment by powerman** [comment]
Well, you get all feedback you can, but at this point you're decision maker, so do whatever you want. Next decision maker will be some Gentoo dev who will review your PR. If he'll decide to merge all these patches in one PR - it's okay to me.

**Comment by clan** [comment]
patches have been applied by gentoo
